mobile/library/cc/stream_callbacks.cc

```diff
@@ -55,52 +64,54 @@ void c_on_trailers(envoy_headers metadata, envoy_stream_intel intel, void* conte
+  StreamCallbacksWrapper* stream_callbacks_wrapper = static_cast<StreamCallbacksWrapper*>(context);
+  StreamCallbacks& stream_callbacks = *stream_callbacks_wrapper->stream_callbacks;
+  if (stream_callbacks.on_error.has_value()) {
     EnvoyErrorSharedPtr error = std::make_shared<EnvoyError>();
     error->error_code = raw_error.error_code;
     error->message = Data::Utility::copyToString(raw_error.message);
     error->attempt_count = absl::optional<int>(raw_error.attempt_count);
-    auto on_error = stream_callbacks->on_error.value();
+    auto on_error = stream_callbacks.on_error.value();
     on_error(error, intel, final_intel);
   }
   release_envoy_error(raw_error);
-  delete stream_callbacks_ptr;
+  delete stream_callbacks_wrapper;
 }
 
 void c_on_complete(envoy_stream_intel intel, envoy_final_stream_intel final_intel, void* context) {
-  auto stream_callbacks_ptr = static_cast<StreamCallbacksSharedPtr*>(context);
-  auto stream_callbacks = *stream_callbacks_ptr;
-  if (stream_callbacks->on_complete.has_value()) {
-    auto on_complete = stream_callbacks->on_complete.value();
+  StreamCallbacksWrapper* stream_callbacks_wrapper = static_cast<StreamCallbacksWrapper*>(context);
+  StreamCallbacks& stream_callbacks = *stream_callbacks_wrapper->stream_callbacks;
+  if (stream_callbacks.on_complete.has_value()) {
+    auto on_complete = stream_callbacks.on_complete.value();
     on_complete(intel, final_intel);
   }
-  delete stream_callbacks_ptr;
+  delete stream_callbacks_wrapper;
 }
 
 void c_on_cancel(envoy_stream_intel intel, envoy_final_stream_intel final_intel, void* context) {
-  auto stream_callbacks_ptr = static_cast<StreamCallbacksSharedPtr*>(context);
-  auto stream_callbacks = *stream_callbacks_ptr;
-  if (stream_callbacks->on_cancel.has_value()) {
-    auto on_cancel = stream_callbacks->on_cancel.value();
+  StreamCallbacksWrapper* stream_callbacks_wrapper = static_cast<StreamCallbacksWrapper*>(context);
+  StreamCallbacks& stream_callbacks = *stream_callbacks_wrapper->stream_callbacks;
+  if (stream_callbacks.on_cancel.has_value()) {
+    auto on_cancel = stream_callbacks.on_cancel.value();
     on_cancel(intel, final_intel);
   }
-  delete stream_callbacks_ptr;
+  delete stream_callbacks_wrapper;
 }
 
 void c_on_send_window_available(envoy_stream_intel intel, void* context) {
-  auto stream_callbacks_ptr = static_cast<StreamCallbacksSharedPtr*>(context);
-  auto stream_callbacks = *stream_callbacks_ptr;
-  if (stream_callbacks->on_send_window_available.has_value()) {
-    auto on_send_window_available = stream_callbacks->on_send_window_available.value();
+  StreamCallbacksWrapper* stream_callbacks_wrapper = static_cast<StreamCallbacksWrapper*>(context);
+  StreamCallbacks& stream_callbacks = *stream_callbacks_wrapper->stream_callbacks;
+  if (stream_callbacks.on_send_window_available.has_value()) {
+    auto on_send_window_available = stream_callbacks.on_send_window_available.value();
     on_send_window_available(intel);
   }
```

Comment:
Make sure to call `delete stream_callbacks_wrapper` here.

Reply:
we don't want to call delete here, because we can still call the other callbacks afterward so need the context.  That's why we only call delete in the `c_on_complete`, `c_on_error`, and `c_on_cancel` functions; because at that point, we won't call the callbacks anymore.

Reply:
Oh I see. That makes sense then.

Reply:
Maybe adding a comment is not a bad idea for this otherwise some people might think we forget to call `delete`.